### DISCUSSION TITLE
`Headers` and `MutableHeaders` incorrectly assumes type of `scope["headers"]`

### DISCUSSION THREAD

**Discussion by jose-rojas0** [encode/starlette#1909] [Potential Issue] [answered]
Looking at the `asgi` docs - https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope

It specifies:
- _headers_ `(Iterable[[byte string, byte string]])`

However Starlette `Headers` assumes it is a list - https://github.com/encode/starlette/blob/master/starlette/datastructures.py#L525 which causes `MutableHeaders` to break when adding a new header (because it wants to use `list.append()` https://github.com/encode/starlette/blob/master/starlette/datastructures.py#L608)

I've found that nginx unit (https://unit.nginx.org/) passes a tuple instead of a list, surfacing this issue.

**Comment by alex-oleshkevich** [comment]
Headers and MutableHeaders serve different purposes. Commonly, Headers is a read only struct that used by request.headers (scope['headers']) and not intended to be modifiable.
MutableHeaders is for responses where you can dynamically alter headers.

The ASGI spec tells that headers are Iterable object (used in for-in like constructs). Starlette conforms this protocol. 

If you want to modify headers in middleware, I suggest converting scope['headers'] to MutableHeaders first.

**Reply by alex-oleshkevich** [reply]
If case of Headers the type of _list is irrelevant. All we need to know is that we can iterate it.

**Reply by adriangb** [reply]
We could make this more explicit with a better variable name and type

**Reply by adriangb** [reply]
Also we should cast it to a list in mutableheaders

**Reply by alex-oleshkevich** [reply]
Yes, it makes sense to convert variable to List at this line
https://github.com/encode/starlette/blob/858629f5188bc79d452600b1eb90eaa0045f6454/starlette/datastructures.py#L525

**Comment by adriangb** [comment, marked as answer]
How does #1917 look?
